aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -492,8 +491,103 @@ impl AggregatorData {
+        self.aggregator_snapshots
+            .insert(snapshot_id, AggregatorSnapshot {
+                id: snapshot_id,
+                state: snapshot_state,
+            });
+        snapshot_id
+    }
+
+    pub fn read_snapshot(
+        &mut self,
+        id: AggregatorID,
+        resolver: &dyn AggregatorResolver,
+    ) -> PartialVMResult<SnapshotValue> {
+        // need to clone here, so we can call self.read_snapshot below.
+        let snapshot_state = match self.aggregator_snapshots.entry(id) {
+            Entry::Vacant(entry) => {
+                // Otherwise, we have to go to storage and read the value.
+                let value_from_storage = resolver
+                    .get_aggregator_v2_value(&id, AggregatorReadMode::Aggregated)
+                    .map_err(|e| {
+                        extension_error(format!(
+                            "Could not find the value of the aggregator: {}",
+                            e
+                        ))
+                    })?;
+                entry
+                    .insert(AggregatorSnapshot {
+                        id,
+                        state: AggregatorSnapshotState::Reference {
+                            speculative_value: SnapshotValue::try_from(value_from_storage)?,
+                        },
+                    })
+                    .state
+                    .clone()
+            },
+            Entry::Occupied(entry) => entry.get().state.clone(),
+        };
+        match snapshot_state {
+            AggregatorSnapshotState::Create { value } => Ok(value),
+            AggregatorSnapshotState::Delta {
+                base_aggregator,
+                delta,
+            } => match self
+                .aggregators
+                .get_mut(&AggregatorVersionedID::V2(base_aggregator))
+            {
+                Some(aggregator) => {
+                    let value = aggregator.read_aggregated_aggregator_value(resolver)?;
+                    Ok(SnapshotValue::Integer(expect_ok(
+                        BoundedMath::new(aggregator.max_value).unsigned_add_delta(value, &delta),
+                    )?))
+                },
+                None => Err(code_invariant_error(
+                    "AggregatorSnapshotState::Delta without base aggregator being set",
+                )),
+            },
+            AggregatorSnapshotState::Derived {
+                base_snapshot,
+                formula,
+            } => {
+                let base = self.read_snapshot(base_snapshot, resolver)?;
+                match base {
+                    SnapshotValue::Integer(v) => Ok(SnapshotValue::String(formula.apply(v))),
+                    SnapshotValue::String(_) => Err(code_invariant_error(
+                        "Tried calling concat on String SnapshotValue",
+                    )),
+                }
+            },
+            AggregatorSnapshotState::Reference { speculative_value } => {
+                Ok(speculative_value.clone())
```

Comment:
don't we own aggregator state (seems it's cloned out)? shouldn't need to clone here?